go/chat/boxer.go

```diff
@@ -622,7 +626,77 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed,
 	}, nil
 }
 
-func (b *Boxer) unboxV2orV3(ctx context.Context, boxed chat1.MessageBoxed,
+func (b *Boxer) validatePairwiseMAC(ctx context.Context, boxed chat1.MessageBoxed) (senderKey []byte, err error) {
+	// First, find a MAC that matches our receiving device encryption KID.
+	ourDeviceKey, err := b.G().ActiveDevice.KeyByType(libkb.DeviceEncryptionKeyType)
+	if err != nil {
+		return nil, err
+	}
+	ourDeviceKeyNacl, ok := ourDeviceKey.(libkb.NaclDHKeyPair)
+	if !ok {
+		return nil, fmt.Errorf("wrong key type, expected NaclDHKeyPair, got %T", ourDeviceKey)
+	}
+	messageMAC, found := boxed.PairwiseMacs[ourDeviceKey.GetKID()]
+	if !found {
+		// This is an error users will actually see when they've just joined a
+		// team or added a new device.
+		return nil, fmt.Errorf("message is not authenticated for this device")
```

Comment:
shoudl we use a new error type for this? https://github.com/keybase/client/blob/master/go/chat/errors.go#L71